### ISSUE TITLE
A hook for creating efficient search index on some languages

### ISSUE THREAD

**Issue by taca** [contao/core#2769]
Hi,

As far as I understand, Contao's serch function assumes that each word are separated by space.  It is true for latin languages but not on some asian languages including Japanese.  I found some codes in system/libraries/Search.php (about line 232 of 55402d485124f58ff8b81257d33d575c), indexPage() function in Search class:

```
        // Split words
        $arrWords = preg_split('/ +/', utf8_strtolower($strText));
        $arrIndex = array();
```

If this preg_split() could relaced with different one by hook function, it would be possible to write the external module which splits words in Japanese efficiently.
(Such an external module needs to use morphological analysis methods and it isn't appropriate to include such function in Contao core.)

This might not enough with this single hook, but I think it would be better results.

Best regards.

--- _Originally created on January 6th, 2011, at 12:09pm (ID 2769)_

**Comment by leofeyer** [comment]
@taca Is this ticket still relevant?

**Comment by taca** [comment]
Yes, it is.
Sadly, http://php.net/manual/en/function.str-word-count.php dose not solve on this problem at all, for no word divider languages.  For these languages, morphological analysis or N-Gram are required to get words.

**Comment by taca** [comment]
And I must say that I can not follow up about implementation of page search on Contao for 4.x.
Still simply use preg_split()?

**Comment by leofeyer** [comment]
Yes.

https://github.com/contao/core-bundle/blob/821041b5ed2dacc4e0aa96cbb0f23c377163fcbd/src/Resources/contao/library/Contao/Search.php#L241

@contao/developers How can we solve this?

**Comment by Toflar** [comment]
Well, even the MySQL Full-Text Search uses whitespaces as delimiters for words. I'm not saying it's good that we do this but also for MySQL it requires [very special handling, installing extensions and running specific queries](https://dev.mysql.com/doc/refman/5.7/en/fulltext-search-mecab.html) to support e.g. Japanese. I'm all in to provide solutions for these type of languages but it's impossible for us to build it, I think. I mean, I absolutely love Katakana but I cannot read nor write it nor so how should I be able to see if results actually make sense based on keywords? 😄

**Comment by ausi** [comment]
The intl extension could help us here to split text into words, e.g.:

```php
function getWords($text, $locale = '') {
    $iterator = \IntlBreakIterator::createWordInstance($locale);
    $iterator->setText($text);
    foreach($iterator->getPartsIterator() as $part) {
        if ($iterator->getRuleStatus() === \IntlBreakIterator::WORD_NONE) {
            continue;
        }
        $words[] = $part;
    }
    return $words;
}

getWords('Das ist ein Test mit Wörtern.', 'de_AT');
// ['Das', 'ist', 'ein', 'Test', 'mit', 'Wörtern']

getWords('セキュリティ上の理由のため、ここでは以下の文字は使用できせん', 'ja_JP');
// ['セキュリティ', '上の', '理由', 'の', 'ため', 'ここ', 'では', '以下', 'の', '文字', 'は', '使用', 'でき', 'せん']
```

**Comment by Toflar** [comment]
That could work although I have no clue if that's enough. E.g. we also highlight the keywords in the search results so there needs to be some handling too I guess (or it needs to be moved). And as I said, no idea if that will provide good results. I'm pretty sure we need a native speaking developer here.

**Comment by leofeyer** [comment]
The keyword highlighting uses the word boundary assertion (`\b`) together with the unicode flag (`/u`), which should handle this case properly.

@ausi Maybe we can even use `\b` to split the string into words?

**Comment by ausi** [comment]
No, the segmentation of text into words is dependent on the locale.  
`\b` does not work for Japanese.

But we could also use the `IntlBreakIterator` to highlight the words. But as @Toflar pointed out, we would need a native speaker that verifies that `IntlBreakIterator` works correctly for this case.

**Comment by leofeyer** [comment]
If we supply an adjusted `Search.php` (maybe as Gist), I'm sure @taca can test if the changes work.

**Comment by leofeyer** [comment]
@ausi Are you going to supply the gist?

**Comment by ausi** [comment]
I think it is enough if @taca checks if the following example was split correctly:

セキュリティ上の理由のため、ここでは以下の文字は使用できせん split into the following words:

- セキュリティ
- 上の
- 理由
- の
- ため
- ここ
- では
- 以下
- の
- 文字
- は
- 使用
- でき
- せん

But I can also provide the gist if desired ☺️.

**Comment by taca** [comment]
How about small fixed sentence? 

セキュリティ上の理由のため、ここでは以下の文字は使用できません。

**Comment by ausi** [comment]
“セキュリティ上の理由のため、ここでは以下の文字は使用できません。” would get split into these words:

- セキュリティ
- 上の
- 理由
- の
- ため
- ここ
- では
- 以下
- の
- 文字
- は
- 使用
- でき
- ま
- せん

**Comment by taca** [comment]
Thanks!

I belive that it produces much better search index for Japanese.  And for more better result,
morphological analysis is required to parse (and there are Open Source libraries for Japanese).
But such morphological analysis methods should be provided as extension.

**Comment by leofeyer** [comment]
Fixed in contao/core-bundle#1629.
